docs: add environment cloning and how-to guide (#2249)
Unleash/unleash: website/docs/user_guide/environments.md

![An API key creation form. The form's fields are "username", "token type", "project", and, crucially, "environment". The development environment is selected.](/img/environments_api_keys.png 'Create Environment specific API Keys')

> ## Cloning environments
> 

Reviewer: We should add the availability note here too.

```suggestion
## Cloning environments

:::note Availability

Environment cloning is an upcoming feature, and is scheduled to become available in one of the next few releases.

:::

```

> If you have a lot of feature toggles, it can be tedious to configure the same activation strategies for each environment. To make this easier, you can clone an environment. This will copy all feature toggle configurations and activation strategies from the source environment to the target environment.
> 
> This can also prove to be a safe way to experiment with activation strategies without affecting production-facing environments. You can clone the production environment, and then test out the new activation strategies in the cloned environment. If everything works as expected, you can then apply the same activation strategies to the production environment.
> 
> When cloning an environment, you can choose several options to customize your new environment: name, type of environment, which projects to clone the environment configuration in, and whether you wish to keep the same user permissions. You'll also be provided the choice of generating a new API key bound to the newly cloned environment right away, or generating one yourself at a later stage.

Reviewer: I'm not sure I understand: the first paragraph reads as if you can choose to copy strategies from an existing environment into another existing environment (e.g. from "dev" to "production"). I might have misunderstood, but I thought this feature only allowed you to create **new** environments that contained the same configuration as the source?

If that's correct, maybe we can be reword this section into something like this:

```suggestion
Unleash environments can be cloned. Cloning an environment creates a **new** environment based on the selected source environment. When cloning an environment, you select any number of projects whose feature toggle configurations will also be cloned. These projects will have identical configurations for the source and target environments. However, the environments can then be configured independently and will not stay in sync with each other.

When cloning an environment, you must give the new environment
- a name
- an environment type
- a list of projects to clone configurations in

You must also decide whether you want to clone the user permissions to the new environment or ... [what do you do here? what's the option?]

Once created, the new environment works just as any other environment.
```

I like the use case suggestions, but I feel like they might be better suited in a "### Use cases" subsection? Wdyt?

> 

Reviewer: We should add a paragraph here explaining how access control works for the newly created enviroments, as @sighphyre mentioned in relation to the how-to guide. This is where we should explain it.

> When cloning an environment, you can choose several options to customize your new environment: Name, type of environment, which projects to clone the environment configuration in, and whether you wish to keep the same user permissions. You'll also be provided the choice of generating a new API key bound to the newly cloned environment right away, or generating one yourself at a later stage.

Reviewer 2: ```suggestion
When cloning an environment, you can choose several options to customize your new environment: name, type of environment, which projects to clone the environment configuration in, and whether you wish to keep the same user permissions. You'll also be provided the choice of generating a new API key bound to the newly cloned environment right away, or generating one yourself at a later stage.
```

![An API key creation form. The form's fields are "username", "token type", "project", and, crucially, "environment". The development environment is selected.](/img/environments_api_keys.png 'Create Environment specific API Keys')

## Cloning environments

:::note Availability

Environment cloning is an upcoming feature, and is scheduled to become available in one of the next few releases.

:::

Unleash environments can be cloned. Cloning an environment creates a **new** environment based on the selected source environment. When cloning an environment, you select any number of projects whose feature toggle configurations will also be cloned. These projects will have identical configurations for the source and target environments. However, the environments can then be configured independently and will not stay in sync with each other.

When cloning an environment, you must give the new environment
- a name
- an environment type

> - a list of projects to clone configurations in

Reviewer: Hmm, maybe this could be a little clearer. It's not obvious when it just says "configurations", I think 😅 

```suggestion
- a list of projects to clone feature configurations in
```

> You must also decide whether you want to clone the user permissions to the new environment or to make it a fresh environment where only an admin or editor of the selected projects can make changes. You can always add permissions later, using [Custom Project Roles](../user_guide/rbac#custom-project-roles).

Reviewer: Nice one! Some minor rewording:

```suggestion
You can also clone user permissions into the new environment. When you do that, permissions in the new environment will be the same as in the environment you cloned from. If you don't clone permissions, only admins and project editors can make changes in the new environment. You can change permissions after creation by using [custom project roles](../user_guide/rbac#custom-project-roles).
```

In order to clone an environment, you can follow the [How to clone environments](../how-to/how-to-clone-environments.mdx) guide.

> Once created, the new environment works just as any other environment.

Author: ```suggestion
Once cloned, the new environment works just as any other environment.
```

Reviewer: I don't agree with this suggestion. In my mind, the **old** environment is **cloned,** the **new** environment is **created**.

Reviewer: I'm not sure I agree with this suggestion. I feel like the "cloned" environment is the source environment (the "clone" being the new one), and that when you clone an environment, you do create a new one regardless.

- Will be possible to turn environments on/off for all projects
- Will be allowed to update and remove environments.
- Will be allowed to create new environments.

> - Will be allowed to clone environments.

Author: Hmm I think I prefer it as a separate item. @thomasheartman WDYT?

> - Will be allowed to clone environments.

Reviewer: Oh, no, timing issues! 😬 I meant that I agree with @sighphyre and that I think it works in a single line. It's just that I wrote the comment before Nuno had followed up, but by the time I'd finished my review, I was too late 🙈 

Regardless though, it doesn't matter much here, so I'm easy either way 😄

- Will get access to two preconfigured environments: “development” and “production”. Existing users of Unleash will also get an additional “default” environment to simplify the adoption of environments.
- Will be possible to turn environments on/off for all projects
- Will be allowed to update and remove environments.

> - Will be allowed to clone or create new environments.

Reviewer: Is it "and" or "or"?  Maybe this makes it clearer?

```suggestion
- Will be allowed to create new environments and clone existing environments.
```
